From 7aa5c3e51ef61ade40c3efda93191640968464cd Mon Sep 17 00:00:00 2001
From: Gautham Kumar Senthilkumaran <quic_gauthamk@quicinc.com>
Date: Thu, 9 Nov 2023 15:03:07 +0530
Subject: [PATCH] mac80211:Fixing the MBSSID crash in driver

Fixing the mbssid beacon length based on mainline. Since there is
difference in the latest code and old mainline code. In latest code it
add extra bits in the beacon which make driver to crash during the
interface bring-up. So, deleting the extra length calculation from
latest code.

Signed-off-by: Gautham Kumar Senthilkumaran <quic_gauthamk@quicinc.com>
Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Signed-off-by: Narendhar Reddy <quic_nare@quicinc.com>
---
 net/mac80211/cfg.c         | 44 ++++++--------------------------------
 net/mac80211/ieee80211_i.h | 12 +----------
 2 files changed, 8 insertions(+), 48 deletions(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 665209e..7d8104a 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1290,7 +1290,7 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 	struct cfg80211_mbssid_elems *mbssid = NULL;
 	struct cfg80211_rnr_elems *rnr = NULL;
 	struct beacon_data *new, *old;
-	int new_head_len, new_tail_len, new_rnr_len = 0;
+	int new_head_len, new_tail_len;
 	int size, err;
 	u64 _changed = BSS_CHANGED_BEACON;
 	struct ieee80211_bss_conf *link_conf = link->conf;
@@ -1325,11 +1325,7 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 			size += struct_size(new->rnr_ies, elem, rnr->cnt);
 		}
 		size += ieee80211_get_mbssid_beacon_len(mbssid, rnr,
-							mbssid->cnt);
-		if (params->rnr_ies) {
-			rnr = params->rnr_ies;
-			new_rnr_len = ieee80211_get_rnr_beacon_len(rnr);
-		}
+												IEEE80211_BCN_EMA_DISABLED);
 	} else if (old && old->mbssid_ies) {
 		mbssid = old->mbssid_ies;
 		size += struct_size(new->mbssid_ies, elem, mbssid->cnt);
@@ -1338,34 +1334,13 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 			size += struct_size(new->rnr_ies, elem, rnr->cnt);
 		}
 		size += ieee80211_get_mbssid_beacon_len(mbssid, rnr,
-							mbssid->cnt);
-		if (old->rnr_ies) {
-			rnr = old->rnr_ies;
-			new_rnr_len = ieee80211_get_rnr_beacon_len(rnr);
+												IEEE80211_BCN_EMA_DISABLED);	
 		}
-	}
 
-	size += new_rnr_len;
 
 	new = kzalloc(size, GFP_KERNEL);
 	if (!new)
 		return -ENOMEM;
-	if (new_rnr_len) {
-		if (params->rnr_ies)
-			new->rnr_ies = kzalloc(struct_size(new->rnr_ies,
-						elem,
-						params->rnr_ies->cnt),
-					GFP_KERNEL);
-		else if (old && old->rnr_ies)
-			new->rnr_ies = kzalloc(struct_size(new->rnr_ies,
-						elem,
-						old->rnr_ies->cnt),
-						GFP_KERNEL);
-		if (!new->rnr_ies) {
-			kfree(new);
-			return -ENOMEM;
-		}
-	}
 	/* start filling the new info now */
 
 	/*
@@ -1384,11 +1359,7 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 		pos += struct_size(new->mbssid_ies, elem, mbssid->cnt);
 		pos += ieee80211_copy_mbssid_beacon(pos, new->mbssid_ies,
 						    mbssid);
-		if (rnr) {
-			new->rnr_ies = (void *)pos;
-			pos += struct_size(new->rnr_ies, elem, rnr->cnt);
-			pos = ieee80211_copy_rnr_beacon(pos, new->rnr_ies, rnr);
-		}
+		pos = ieee80211_copy_rnr_beacon(pos, new->rnr_ies, rnr);
 		/* update bssid_indicator */
 		link_conf->bssid_indicator =
 			ilog2(__roundup_pow_of_two(mbssid->cnt + 1));
@@ -1420,7 +1391,6 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 	err = ieee80211_set_probe_resp(sdata, params->probe_resp,
 				       params->probe_resp_len, csa, cca, link);
 	if (err < 0) {
-		kfree(new->rnr_ies);
 		kfree(new);
 		return err;
 	}
@@ -1437,7 +1407,6 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 							 link_conf);
 
 		if (err < 0) {
-			kfree(new->rnr_ies);
 			kfree(new);
 			return err;
 		}
@@ -1454,8 +1423,9 @@ ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 	}
 
 	*changed |= _changed;
-	return 0;
-}
+	return _changed;
+
+	}
 
 static int ieee80211_start_ap(struct wiphy *wiphy, struct net_device *dev,
 			      struct cfg80211_ap_settings *params)
diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index bf5c225..e8b76ad 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1214,7 +1214,7 @@ ieee80211_vif_get_shift(struct ieee80211_vif *vif)
 static inline int
 ieee80211_get_mbssid_beacon_len(struct cfg80211_mbssid_elems *elems,
 				struct cfg80211_rnr_elems *rnr_elems,
-				u8 i)
+				int i)
 {
 	int len = 0;
 
@@ -1224,16 +1224,6 @@ ieee80211_get_mbssid_beacon_len(struct cfg80211_mbssid_elems *elems,
 	if (i >= IEEE80211_BCN_EMA_BASE)
 		return elems->elem[i].len;
 
-	if (i < elems->cnt) {
-		len = elems->elem[i].len;
-		if (rnr_elems) {
-			len += rnr_elems->elem[i].len;
-			for (i = elems->cnt; i < rnr_elems->cnt; i++)
-				len += rnr_elems->elem[i].len;
-		}
-		return len;
-	}
-
 	/* i == elems->cnt, calculate total length of all MBSSID elements */
 	for (i = 0; i < elems->cnt; i++)
 		len += elems->elem[i].len;
-- 
2.17.1

